File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/features/cheats/model/ARCheat.java
// SPDX-License-Identifier: GPL-2.0-or-later

package org.dolphinemu.dolphinemu.features.cheats.model;

import androidx.annotation.Keep;
import androidx.annotation.NonNull;

public class ARCheat extends AbstractCheat
{
  @Keep
  private final long mPointer;

  public ARCheat()
  {
    mPointer = createNew();
  }

  @Keep
  private ARCheat(long pointer)
  {
    mPointer = pointer;
  }

  @Override
  public native void finalize();

  private native long createNew();

  public boolean supportsCreator()
  {
    return false;
  }

  public boolean supportsNotes()
  {
    return false;
  }

  @NonNull
  public native String getName();

  @NonNull
  public native String getCode();

  public native boolean getUserDefined();

  public native boolean getEnabled();

  @Override
  protected native int trySetImpl(@NonNull String name, @NonNull String creator,
          @NonNull String notes, @NonNull String code);

  @Override
  protected native void setEnabledImpl(boolean enabled);

  @NonNull
  public static native ARCheat[] loadCodes(String gameId, int revision);

  public static native void saveCodes(String gameId, int revision, ARCheat[] codes);
}

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/features/cheats/model/ARCheat.kt
// SPDX-License-Identifier: GPL-2.0-or-later

package org.dolphinemu.dolphinemu.features.cheats.model

import androidx.annotation.Keep

class ARCheat : AbstractCheat {
    @Keep
    private val pointer: Long

    constructor() {
        pointer = createNew()
    }

    @Keep
    private constructor(pointer: Long) {
        this.pointer = pointer
    }

    external fun finalize()

    private external fun createNew(): Long

    override fun supportsCreator(): Boolean = false

    override fun supportsNotes(): Boolean = false

    external override fun getName(): String

    external override fun getCode(): String

    external override fun getUserDefined(): Boolean

    external override fun getEnabled(): Boolean

    external override fun setEnabledImpl(enabled: Boolean)

    external override fun setCheatImpl(
        name: String,
        creator: String,
        notes: String,
        code: String
    ): Int

    companion object {
        @JvmStatic
        external fun loadCodes(gameId: String, revision: Int): Array<ARCheat>

        @JvmStatic
        external fun saveCodes(gameId: String, revision: Int, codes: Array<ARCheat>)
    }
}
